### PR TITLE
Ghost-role racial bans

### LINKED ISSUE #905: Ashwalker can join dauntless
## Issue Summary

See title

## Round ID:

N/A

## Reproduction:

Join as ashwalker on a dauntless ghost spawner

### PR DESCRIPTION
**Pull request by aKromatopzia**
## About The Pull Request

Tested only for Tarkon and Dauntless. Functional. Does the same for DS2, black market dealer, space hoteliers, etc.

resolves #905 
## Why It's Good For The Game

Simply makes it so that you can't join Tarkon, Dauntless, Interdyne, DS2, or space hoteliers as an ashwalker or icedweller.

## Changelog

:cl:
code: banned_species var, which is an inverse of restricted_species in purpose.
code: you can't join certain ghost roles with ashwalker or icedwellers anymore.
/:cl:

### CODE REVIEW

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm [hunk 1]
```diff
@@ -12,11 +12,12 @@
 	icon_state = "sleeper_s"
 	prompt_name = "Dauntless Personnel"
 	you_are_text = "You are a syndicate Operative, employed on a top secret spy vessel hidden on lavaland"
-	flavour_text = "Unfortinately, your hated enemy, NanoTrasen, has started mining in this sector. Continue working the best you can, and keep a low profile"
+	flavour_text = "Unfortunately, your hated enemy, NanoTrasen, has started mining in this sector. Continue working the best you can, and keep a low profile" //fixed typo
```
Comment (Majkl-J) [inline modular_zubbers/maps/offstation/dauntless/mob_spawns.dm:15]: Fix the typo upstream

File: code/modules/mob_spawn/mob_spawn.dm [hunk 2]
```diff
@@ -151,6 +151,9 @@
 	/// Are we limited to a certain species type? LISTED TYPE
 	var/restricted_species
 	// SKYRAT EDIT END
+	//BUBBER EDIT
+	var/banned_species
```
Comment (Majkl-J) [inline code/modules/mob_spawn/mob_spawn.dm:155]:
```suggestion
	var/list/banned_species
```

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm [hunk 3]
```diff
@@ -12,11 +12,12 @@
...
 	quirks_enabled = TRUE
 	random_appearance = FALSE
 	loadout_enabled = TRUE
 	computer_area = /area/ruin/space/has_grav/bubbers/dauntless/service/dorms
+	banned_species = list(/datum/species/human/felinid/primitive, /datum/species/lizard/ashwalker) //BUBBER EDIT added species restriction for icecat and ashliz
```
Comment (Majkl-J) [inline modular_zubbers/maps/offstation/dauntless/mob_spawns.dm:20]: Probably would be better off making these limits apply to parent and simply override the list with an empty one on the cafe, icemoon hole, and ashie egg spawners

**Review by Majkl-J — CHANGES_REQUESTED** [review]

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm (hunk 3 above)
Comment (Majkl-J) [reply modular_zubbers/maps/offstation/dauntless/mob_spawns.dm:20]: Also these can be modular

Comment (aKromatopzia) [reply modular_zubbers/maps/offstation/dauntless/mob_spawns.dm:20]: How could these be modular if I'd have to go into each to add the var anyways?

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm (hunk 1 above)
Comment (StrangeWeirdKitten) [reply modular_zubbers/maps/offstation/dauntless/mob_spawns.dm:15]: Hold off on that

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm (hunk 3 above)
Comment (StrangeWeirdKitten) [reply modular_zubbers/maps/offstation/dauntless/mob_spawns.dm:20]: Dauntless code is already modular.

**Comment by aKromatopzia** [comment]
is . - ..() the supercall thing or

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm (hunk 1 above)
Comment (Majkl-J) [reply modular_zubbers/maps/offstation/dauntless/mob_spawns.dm]: Oh bruh I didn't notice

File: modular_zubbers/maps/offstation/dauntless/mob_spawns.dm (hunk 3 above)
Comment (Majkl-J) [reply modular_zubbers/maps/offstation/dauntless/mob_spawns.dm]: > How could these be modular if I'd have to go into each to add the var anyways?

Sorry I meant the edits in general I was reviewing this at 1 AM and it shows

File: modular_zubbers/maps/offstation/mob_spawns.dm [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/obj/effect/mob_spawn/ghost_role/human/oldeng
+	banned_species = list(/datum/species/human/felinid/primitive, /datum/species/lizard/ashwalker)
+/obj/effect/mob_spawn/ghost_role/human/oldsci
+	banned_species = list(/datum/species/human/felinid/primitive, /datum/species/lizard/ashwalker)
+/obj/effect/mob_spawn/ghost_role/human/oldsec
+	banned_species = list(/datum/species/human/felinid/primitive, /datum/species/lizard/ashwalker)
+
+//various bubber roles
+/obj/effect/mob_spawn/ghost_role/human/dauntless
+	banned_species = list(/datum/species/human/felinid/primitive, /datum/species/lizard/ashwalker)
```
Comment (Majkl-J) [inline modular_zubbers/maps/offstation/mob_spawns.dm:37]:
```suggestion
/obj/effect/mob_spawn/ghost_role/human
	banned_species = list(/datum/species/human/felinid/primitive, /datum/species/lizard/ashwalker)
```

File: modular_zubbers/maps/offstation/mob_spawns.dm [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/obj/effect/mob_spawn/ghost_role
+	var/banned_species //functions similarly to restricted_species, but is instead a list of what NOT to allow.
+	
+/obj/effect/mob_spawn/ghost_role/attack_ghost(mob/dead/observer/user)
+	if(banned_species && (user.client?.prefs?.read_preference(/datum/preference/choiced/species) in banned_species))
```
Comment (Majkl-J) [inline modular_zubbers/maps/offstation/mob_spawns.dm:5]:
```suggestion
	if(!restricted_species && banned_species && (user.client?.prefs?.read_preference(/datum/preference/choiced/species) in banned_species))
```

**Review by Majkl-J — CHANGES_REQUESTED** [review]
Apologies for previous half assed review, lesson learned about reviewing at 1AM.

Anyhow I thought about this and we probably don't need this check to run together with restricted species, since their uses are mutually exclusive, and removed the line of overrides with only one on the parent.

Also gonna ping @StrangeWeirdKitten so he can double check this suggestion before you commit

File: modular_zubbers/maps/offstation/mob_spawns.dm (hunk 4 above)
Comment (aKromatopzia) [reply modular_zubbers/maps/offstation/mob_spawns.dm:37]: No, can't do this. Ash walkers are obj/effect/mob_spawn/ghost_role/human/ash_walker

**Comment by aKromatopzia** [comment]
> Apologies for previous half assed review, lesson learned about reviewing at 1AM.
> 
> Anyhow I thought about this and we probably don't need this check to run together with restricted species, since their uses are mutually exclusive, and removed the line of overrides with only one on the parent.
> 
> Also gonna ping @StrangeWeirdKitten so he can double check this suggestion before you commit

I think the changes I just made however should work? varname = "" does blank out a var iirc

**Review by Majkl-J — APPROVED** [review]
Fabled green checkmark applied. Do mind for next time `"null"` is different from `null`, otherwise all good!

**Pull request merged**